crates/bevy_ecs/src/query/state.rs

```diff
@@ -40,6 +39,27 @@ impl<Q: WorldQuery, F: WorldQuery> FromWorld for QueryState<Q, F> {
     }
 }
 
+impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
+    pub(crate) fn as_immut(&self) -> &QueryState<Q::ReadOnly, F::ReadOnly> {
+        // Safety: invariant on `WorldQuery` trait upholds that `Q::ReadOnly` and `F::ReadOnly`
+        // have a subset of the access, and match the same archetypes/tables as `Q`/`F` respectively.
+        unsafe { self.as_whatever::<Q::ReadOnly, F::ReadOnly>() }
+    }
+
+    /// # Safety
+    ///
+    /// `NewQ` must have a subset of the access that `Q` does and match the same archetypes/tables
+    /// `NewF` must have a subset of the access that `F` does and match the same archetypes/tables
+    pub(crate) unsafe fn as_whatever<
```

Comment:
We probably need a better name for this.

Reply:
I agree but its `pub(crate)` so it doesnt really matter :thinking:

Reply:
```suggestion
    pub(crate) unsafe fn transmute_query<
```

Reply:
i want `as` in there somewhere because i'd expect us to add `into_whatever` or `into_immut` into the future

Reply:
`as_transmuted_query`?  `as_subquery`?

Just `as_query` is a little too innocent for my liking.